### PR TITLE
Improved arrow functionality.

### PR DESCRIPTION
**Pull request by KingLemming**
This patch basically allows for an itemstack which is an "arrow" to determine if it is indeed valid, rather than a hard check for instanceof ItemArrow.

The use case for this is a Quiver, which would "look" like an arrow by returning true when it had arrows, and false when it did not.

Existing bow logic is already sufficient to prevent the Quiver from being deleted (isInfinite) and creating the correct arrows (createArrow).

### CODE REVIEW

**Comment by KingLemming** [comment]
Just going to add a comment as to WHY this is an improvement.

The arrow nock / loose events do not properly address what this PR changes - the ability for an ItemStack to temporarily be ignored by findAmmo(). The events can't actually do this. They can prevent the ammo from being nocked in the first place, but that halts the process outright. But for illustration, here's how that would work:

1. Registering a new event bus handler.
2. Catching the arrow nock event.
3. Trying to find ammo using the ItemStack provided - which is the BOW.*, **
4. Returning a fail if the first ItemStack found should be considered "invalid."***

*This will fail outright if the arrow nock event is fired from something that doesn't extend ItemBow.
** This also assumes that the bow uses the standard method of finding ammo - literally the findAmmo() method inside ItemBow.
***This isn't what we want - this means that if the first "ammo" found is "invalid," then the bow can't be nocked.

Notice point *** there - that is the critical difference - event handling does not in any way provide a way for the bow to **ignore** ammo.

Here's the use case for that: Bow in slot 0, Empty Quiver in Slot 1, stack of Arrows in slot 2.

Just for clarity's sake, a Quiver is an Item which holds stacks of arrows. It extends ItemArrow (clunky but unobtrusive) and when told to create an arrow, uses a stored stack to create the EntityArrow object, then decrements that stack internally.

With this implementation, the quiver will return false on an isValidArrow() call, meaning that the bow will then keep searching and find the arrows in slot 2.

If we tried to handle this through events instead, the bow would find the quiver, we'd have to do all of the event handling, and tell to bow to go pound sand, canceling the nock outright and not allowing the arrows in slot 2 to be considered as valid ammo.

So basically, this is the cleanest way to handle this situation without completely rewriting the arrow nock and loose events. It's backwards compatible, breaks nothing at all, and is unobtrusive.

**Comment by KingLemming** [comment]
After giving this more thought, while it doesn't harm anything, I'm going to shut it down and handle it rather inefficiently through other means.
